### PR TITLE
StateManager and LoadBlockIndex

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/blockchain/blocktree.hpp [hunk 1]
```diff
@@ -86,7 +86,8 @@ struct BlockTree {
     // our store (yet) to check it correctly
     for (size_t i = 1, size = chain.size(); i < size; i++) {
       auto& block = chain[i];
-      if (!this->acceptBlock(block, state, false)) {
+      index_t temp_index;
+      if (!this->acceptBlock(block, temp_index, state, false)) {
         return state.addStackFunction("bootstrapWithChain()");
       }
```
Comment (blackyblack) [inline include/veriblock/blockchain/blocktree.hpp]: Should explicitly ignore temp_index, eg (void)temp_index;

File: include/veriblock/blockchain/blocktree.hpp [hunk 2]
```diff
@@ -299,13 +305,13 @@ struct BlockTree {
       return state.addStackFunction("acceptBlock");
     }
 
-    auto* index = insertBlockHeader(block);
+    auto index = insertBlockHeader(block);
```
Comment (blackyblack) [inline include/veriblock/blockchain/blocktree.hpp]: You can write directly to blockIndex, eg &blockIndex = insertBlockHeader(block);

File: include/veriblock/state_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+#include <vector>
+
+#include "veriblock/blockchain/blocktree.hpp"
+#include "veriblock/state_manager.hpp"
+#include "veriblock/validation_state.hpp"
+
+namespace VeriBlock {
+
+template <typename Block_t, typename ChainParams_t>
+bool LoadBlockTree(
```
Comment (blackyblack) [inline include/veriblock/state_utils.hpp]: Use camelCase, eg loadBlockTree

Comment (blackyblack) [reply include/veriblock/state_utils.hpp]: Not clear what is being loaded here. Please be more specific.

Comment (blackyblack) [reply include/veriblock/state_utils.hpp]: Maybe this function should belong to some State or Manager class.

File: src/mock_miner.cpp [hunk 4]
```diff
@@ -20,12 +20,13 @@ bool mineBlocks(const uint32_t& n,
...
                 ValidationState& state) {
   bool res = true;
+  BlockIndex<Block> temp_index;
   for (uint32_t i = 0; i < n && res; ++i) {
     BlockIndex<Block>* tip = blockChain->getBestChain().tip();
     assert(tip != nullptr);
 
     Block block = miner->createNextBlock(*tip, {});
-    res = blockChain->acceptBlock(block, state);
+    res = blockChain->acceptBlock(block, temp_index, state);
```
Comment (blackyblack) [inline src/mock_miner.cpp]: Move BlockIndex<Block> temp_index inside the loop - decrease it's scope.
Explicitly ignore result, eg, (void)temp_index;

File: src/state_manager.cpp [hunk 5]
```diff
@@ -1 +1,16 @@
...
+  btcRepoBatch->put(block);
+}
+
+void StateChange::putVbkBlock(const BlockIndex<VbkBlock>& block) {
+  vbkRepoBatch->put(block);
+}
+
+void StateChange::commit() {
+  btcRepoBatch->commit();
+  vbkRepoBatch->commit();
```
Comment (blackyblack) [inline src/state_manager.cpp]: Probably some error checking and rollbacks should be here.

**Comment by Warchant** [comment]
Add methods `addPayloads` and `removePayloads` to `StateChange`. Doing this way, we can omit coupling of PopManager & StateManager

File: test/state_manager_test.cpp [hunk 6]
```diff
@@ -1,8 +1,137 @@
...
+          *expectedVbkBlockTree->getBestChain().tip(), {});
+
+      ASSERT_TRUE(expectedVbkBlockTree->acceptBlock(
+          newVbkBlock, state, &temp_index_vbk));
+      ASSERT_TRUE(state.IsValid());
+      change.putVbkBlock(temp_index_vbk);
+    }
+
+    change.commit();
+    stateManager.flush();
```
Comment (blackyblack) [inline test/state_manager_test.cpp]: Move stateManager.flush() inside change.commit().

File: include/veriblock/state_utils.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  std::sort(blocks.begin(),
+            blocks.end(),
+            [](const BlockIndex<Block_t>& block1,
+               const BlockIndex<Block_t>& block2) -> bool {
+              return block1.height < block2.height;
+            });
+
+  for (const auto& block : blocks) {
+    if (!tree.acceptBlock(block.header, state)) {
+      return state.addStackFunction("LoadBlockTree()");
```
Comment (blackyblack) [inline include/veriblock/state_utils.hpp]: loadBlockTree

**Pull request merged**